lib/finch/http1/conn.ex

```diff
@@ -169,44 +169,73 @@ defmodule Finch.Conn do
 
   defp handle_response(response, conn, metadata, start_time, extra_measurements) do
     case response do
-      {:ok, mint, {status, headers, _} = acc} ->
+      {:ok, mint, acc, {status, headers}} ->
         metadata = Map.merge(metadata, %{status: status, headers: headers})
         Telemetry.stop(:recv, start_time, metadata, extra_measurements)
         {:ok, %{conn | mint: mint}, acc}
 
-      {:error, mint, error} ->
-        metadata = Map.put(metadata, :error, error)
+      {:error, mint, error, {status, headers}} ->
+        metadata = Map.merge(metadata, %{error: error, status: status, headers: headers})
         Telemetry.stop(:recv, start_time, metadata, extra_measurements)
         {:error, %{conn | mint: mint}, error}
     end
   end
 
-  defp receive_response([], acc, fun, mint, ref, timeout) do
+  defp receive_response(entries, acc, fun, mint, ref, timeout, status \\ nil, headers \\ [])
+
+  defp receive_response([], acc, fun, mint, ref, timeout, status, headers) do
     case HTTP1.recv(mint, 0, timeout) do
       {:ok, mint, entries} ->
-        receive_response(entries, acc, fun, mint, ref, timeout)
+        receive_response(entries, acc, fun, mint, ref, timeout, status, headers)
 
       {:error, mint, error, _responses} ->
-        {:error, mint, error}
+        {:error, mint, error, {status, headers}}
     end
   end
 
-  defp receive_response([entry | entries], acc, fun, mint, ref, timeout) do
+  defp receive_response([entry | entries], acc, fun, mint, ref, timeout, status, headers) do
     case entry do
       {:status, ^ref, value} ->
-        receive_response(entries, fun.({:status, value}, acc), fun, mint, ref, timeout)
+        receive_response(
+          entries,
+          fun.({:status, value}, acc),
+          fun,
+          mint,
+          ref,
+          timeout,
+          value,
+          headers
+        )
 
       {:headers, ^ref, value} ->
-        receive_response(entries, fun.({:headers, value}, acc), fun, mint, ref, timeout)
+        receive_response(
+          entries,
+          fun.({:headers, value}, acc),
+          fun,
+          mint,
+          ref,
+          timeout,
+          status,
+          headers ++ value
```

Comment:
I'm not familiar with the mint internals but can this message received more than once? If not then this can be just `value` right? (instead of `headers ++ value`)

Reply:
I believe that we could see multiple `{:headers, ...}` tuples if a response had trailing headers in addition to standard headers. I am looking at how I could write a test case for that now...

Reply:
It appears that while Cowboy does support trailing headers, Plug.Cowboy does not yet allow you to send them.

Reply:
Ahh too bad..